ir/src/ir2/graph/nodes.rs

```diff
@@ -0,0 +1,236 @@
+            NodeTypes::Function(function) => function.get_children(),
+        };
+        children
+    }
+    fn remove_child(&mut self, child: Link<NodeTypes>) -> Link<NodeTypes> {
+        println!("Link<NodeTypes>.remove_child({:?})", child);
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Graph(graph) => graph.remove_child(child),
+            NodeTypes::Node(node) => node.remove_child(child),
+            NodeTypes::I32(_) => panic!("remove_child called on leaf"),
+            NodeTypes::Add(add) => add.remove_child(child),
+            NodeTypes::Function(function) => function.body().remove_child(child),
+        }
+    }
+    fn first(&self) -> Link<NodeTypes> {
+        match self.borrow().deref() {
+            NodeTypes::Graph(graph) => graph.first(),
+            NodeTypes::Node(node) => node.first(),
+            NodeTypes::I32(_) => panic!("first() called on leaf"),
+            NodeTypes::Add(add) => add.first(),
+            NodeTypes::Function(function) => function.body().first(),
+        }
+    }
+    fn last(&self) -> Link<NodeTypes> {
+        match self.borrow().deref() {
+            NodeTypes::Graph(graph) => graph.last(),
+            NodeTypes::Node(node) => node.last(),
+            NodeTypes::I32(_) => panic!("last() called on leaf"),
+            NodeTypes::Add(add) => add.last(),
+            NodeTypes::Function(function) => function.body().last(),
+        }
+    }
+    fn new_i32(&mut self, data: i32) -> Link<NodeTypes> {
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Graph(graph) => graph.new_i32(data),
+            NodeTypes::Node(node) => node.new_i32(data),
+            NodeTypes::I32(_) => panic!("new_i32() called on leaf"),
+            NodeTypes::Add(add) => add.new_i32(data),
+            NodeTypes::Function(function) => function.body().new_i32(data),
+        }
+    }
+    fn new_add(&mut self) -> Link<NodeTypes> {
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Graph(graph) => graph.new_add(),
+            NodeTypes::Node(node) => node.new_add(),
+            NodeTypes::I32(_) => panic!("new_add() called on leaf"),
+            NodeTypes::Add(add) => add.new_add(),
+            NodeTypes::Function(function) => function.new_add(),
+        }
+    }
+    fn new_body(&mut self) -> Link<NodeTypes> {
+        match self.borrow_mut().deref_mut() {
+            NodeTypes::Graph(graph) => graph.new_body(),
+            NodeTypes::Node(_) => panic!("new_body() called on body"),
+            NodeTypes::I32(_) => panic!("new_body() called on leaf"),
+            NodeTypes::Add(_) => panic!("new_body() called on add"),
+            NodeTypes::Function(_) => panic!("new_body() called on function"),
+        }
+    }
+}
+
+impl Child for Link<NodeTypes> {
+    fn get_parent(&self) -> BackLink<NodeTypes> {
+        match self.borrow().deref() {
+            NodeTypes::Graph(_) => panic!("get_parent called on graph"),
```

Comment:
I assume the goal of this NodeTypes is to represent nodes that don't have a parent.
IMO it's clearer to have a NodeTypes::Root that only reference one "Node", and then a Graph is basically a `Vec<Root>` (or a `Graph { boundary_roots: Vec<Root>, integrity_roots: Vec<Root> }`).

Would that be possible? Or does this require too much change / is not possible because of some edge case?

The main reason not to do it I think is if this duplicates code to handle adding roots to a graph or similar

Reply:
It is possible, but that currently means duplicating the code almost verbatim for the graph builder pattern (modulo what's generic and what's behind the enum variant)

I was thinking of splitting NodeType(s) into multiple categories, but havent yet found a way to make that generic enough to avoid big code duplication. If I find a way to, I'll make it a separate one with just the graph. For now, this has been the easiest to implement/extend upon.